tidy3d/plugins/mode/mode_solver.py

```diff
@@ -125,110 +87,148 @@ def solve(self, mode_spec: ModeSpec) -> List[ModeInfo]:
 
-        eps_xx = self.simulation.epsilon(plane, "Ex", self.freq)
-        eps_yy = self.simulation.epsilon(plane, "Ey", self.freq)
-        eps_zz = self.simulation.epsilon(plane, "Ez", self.freq)
+        # swap axes to plane coordinates (normal_axis goes to z)
+        eps_sim_ax = (eps_xx, eps_yy, eps_zz)
+        eps_zz, (eps_xx, eps_yy) = self.plane.pop_axis(eps_sim_ax, axis=self.normal_axis)
 
+        # construct eps to feed to mode solver
         return np.stack((eps_xx, eps_yy, eps_zz), axis=0)
 
+    def rotate_field_coords(self, field):
+        """move the propagation axis=z to the proper order in the array"""
+        f_x, f_y, f_z = np.moveaxis(field, source=3, destination=1 + self.normal_axis)
+        f_rot = np.stack(self.plane.unpop_axis(f_z, (f_x, f_y), axis=self.normal_axis), axis=0)
+        return f_rot
+
+    def process_fields(
+        self, mode_fields: Array[complex], freq_index: int, mode_index: int
+    ) -> Tuple[FIELD, FIELD]:
+        """transform solver fields to simulation axes, set gauge, and check decay at boundaries."""
+
+        # Separate E and H fields (in solver coordinates)
+        E, H = mode_fields[..., mode_index]
+
+        # Warn if not decayed at edges
+        e_edge = 0
+        if E.shape[1] > 1:
+            e_edge = np.sum(np.abs(E[:, 0, :]) ** 2 + np.abs(E[:, -1, :]) ** 2)
+        if E.shape[2] > 1:
+            e_edge += np.sum(np.abs(E[:, :, 0]) ** 2 + np.abs(E[:, :, -1]) ** 2)
+        e_norm = np.sum(np.abs(E) ** 2)
+
+        if e_edge / e_norm > FIELD_DECAY_CUTOFF:
+            logging.warning(
+                f"Mode field at frequency index {freq_index}, mode index {mode_index} does not "
+                "decay at the plane boundaries."
+            )
+
+        # Set gauge to highest-amplitude in-plane E being real and positive
+        ind_max = np.argmax(np.abs(E[:2]))
+        phi = np.angle(E[:2].ravel()[ind_max])
+        E *= np.exp(-1j * phi)
+        H *= np.exp(-1j * phi)
+
+        # Rotate back to original coordinates
+        (Ex, Ey, Ez) = self.rotate_field_coords(E)
+        (Hx, Hy, Hz) = self.rotate_field_coords(H)
+
+        # apply -1 to H fields if a reflection was involved in the rotation
+        if self.normal_axis == 1:
+            Hx *= -1
+            Hy *= -1
+            Hz *= -1
+
+        return ((Ex, Ey, Ez), (Hx, Hy, Hz))
+
     def to_source(
-        self, mode_spec: ModeSpec, fwidth: float, direction: Direction, mode_index: int = 0
+        self,
+        mode_spec: ModeSpec,
+        source_time: SourceTime,
+        direction: Direction,
+        mode_index: int = 0,
     ) -> ModeSource:
         """Creates ``ModeSource`` from a Mode + additional specifications.
 
         Parameters
         ----------
         mode_spec : ModeSpec
             :class:`ModeSpec` object containing specifications of mode.
-        fwidth : float
-            Standard deviation of ``GaussianPulse`` of source (Hz).
+        source_time: SourceTime
```

Comment:
probably should use `:class:` stuff in here for source time

Reply:
So same for ModeSpec in `mode_spec : ModeSpec` then?